### ISSUE TITLE
License Issue

### ISSUE THREAD

**Issue by LukasHindemith** [SebastianJanampa/DETRPose#7]
Hey,

thanks for the awesome work. While reading through the files, I recognized that you include Ultralytics library for YOLO benchmarking. As Ultralytics applies the AGPL-3.0 license, to my knowledge all derivatives that include AGPL-3.0 code also have to apply the same license. For easier usability of the community, maybe you could consider to exclude the Ultralytics library?

Best
Lukas

**Comment by SebastianJanampa** [comment]
Hi, 
Thanks for your interest in our work. Could you please provide more information on which files? 
I am relatively new to licensing, and that's something I don't quite understand well. I know I modified the Ultralytics code for visualization, but not for benchmarking. In fact, I do not require installing the Ultralytics library to run my code. Anyways, please point out the files I should modify; I would be more than happy to do it for the computer vision community.

**Comment by LukasHindemith** [comment]
Hey,

you import the ultralytics library in the following file:
https://github.com/SebastianJanampa/DETRPose/blob/main/tools/deployment/export_yolo_w_nms.py
Also, you reference to ultraltics for the annotator color palette.
I would suggest to exclude the direct import of ultralytics. In that way, your code can be licensed with Apache2.0.

**Comment by davidpagnon** [comment]
This is not completed yet :)

If you could get rid of all references to Ultralytics (see [here](https://github.com/search?q=repo%3ASebastianJanampa%2FDETRPose%20ultralytics&type=code)), you could legally use an Apache license instead of abiding by the Ultralytics one which is quite expensive for business, and you would reach a much wider audience. Of course, you cannot just copy-paste the functions; you'd have to rewrite them.

I'm particularly interested because I'm considering integrating DETRPose in [Sports2D](https://github.com/davidpagnon/Sports2D) and [Pose2Sim](https://github.com/perfanalytics/pose2sim), which have a free license that is not compatible with the Ultralytics one.

Note that I would also need DETRPose to support feet keypoints such as in the *Halpe26* and/or *Coco133* datasets (see how RTMPose does [here](https://github.com/open-mmlab/mmpose/blob/main/projects/rtmpose/README.md#26-keypoints)), but that's another topic that I'm willing to discuss with you if you are interested.

**Comment by SebastianJanampa** [comment]
Hello,
Thank you for your interest in DETRPose. 
I will work on the visualization this weekend.  


> Note that I would also need DETRPose to support feet keypoints, such as in the _Halpe26_ and/or _Coco133_ datasets (see how RTMPose does [here](https://github.com/open-mmlab/mmpose/blob/main/projects/rtmpose/README.md#26-keypoints)), but that's another topic that I'm willing to discuss with you if you are interested.

This is an interesting idea. However, DETRPose may not be the best option to train on those datasets because the GroupAttention complexity is O(k^2), where k is the number of keypoints. I have some ideas in mind to improve DETRPose; however, I haven't had time to develop this new model.

**Comment by SebastianJanampa** [comment]
Hi @LukasHindemith  and @davidpagnon 

I just want to let you know I've updated the inference code. I completely got rid of ultralytics for the visualization. I hope this update helps you in your projects. 

Merry Christmas and Happy New Year :)

**Comment by davidpagnon** [comment]
That was fast, thanks!

Thanks for explaining to me the limitations when it comes to the number of keypoints, I did not realize that. I suppose that it could become problematic for Coco133, but not so much for Halpe26, right? This is the only one I really need.

**Comment by SebastianJanampa** [comment]
That's right. The other problem is the vRAM. I noticed that I can run D-FINE-L with a batch size of 16 on my 24GB GPU without any issues. On the other hand, for DETRPose-L, I had to use a batch size of 8 with a gradient accumulation of 2, to be able to use a batch size of 16. However, there's an issue with using gradient accumulation: the batchnorm layer, used in the backbone and encoder. For the final results, I got help from Lambda.ai, which let me use 8 Tesla V100 for distributed training (I only needed to use 4 of the 8).

If you prefer YOLO-based models, I would recommend [RTMO](https://github.com/open-mmlab/mmpose/tree/main/projects/rtmo) (which has Apache 2.0 license, and it also predicts bounding boxes). I would also like to say that DETRPose was developed because I couldn't install mmpose due to issues with installing the mimcv library. Nevertheless, they provide the ONNX files (I am not sure if they are easy to export to TensorRT, since the documentation says you have to build some custom operations), and you can use them through the [RTMlib](https://github.com/Tau-J/rtmlib/tree/main/rtmlib) library.

**Comment by davidpagnon** [comment]
I know about RTMO, it is actually made available within [RTMlib](https://github.com/Tau-J/rtmlib) without the need for MMPose :)

I'm not sure I understand all the technical points above, but I thought DETR-pose had a few advantages over RTMO?

Edit: I misread your comment, you know about RTMlib

**Comment by SebastianJanampa** [comment]
> I'm not sure I understand all the technical points above, but I thought DETR-pose had a few advantages over RTMO?

Yes, it does. The main advantages are end-to-end inference and faster convergence. In terms of AP, both models have similar results.

**Comment by davidpagnon** [comment]
Thanks! (And sorry for the completely off topic discussion 😁)
